File: mg_slam/scripts/slam_gnss_2d/optimizer/gtsam_optimizer.py
from __future__ import annotations

import logging
from typing import Sequence

import numpy as np
from gtsam import (
    BetweenFactorPose2,
    LevenbergMarquardtOptimizer,
    LevenbergMarquardtParams,
    NonlinearFactorGraph,
    Pose2,
    PoseTranslationPrior2D,
    PriorFactorPose2,
    Values,
    noiseModel,
)


from slam_gnss_2d.core.data_types import GnssPrior, PoseEdge, PoseNode

_logger = logging.getLogger(__name__)

# 最初のノードを固定するアンカー拘束の分散値 (x, y, yaw)
# 非常に小さい値で最初のノードを強く固定する
_ANCHOR_VARIANCES = np.array([1e-6, 1e-6, 1e-8])


class GTSAMOptimizer:
    """GTSAM LevenbergMarquardt による 2D ポーズグラフ最適化。"""

    def optimize(
        self,
        nodes: list[PoseNode],
        edges: list[PoseEdge],
        gnss_priors: Sequence[GnssPrior] = (),
    ) -> list[PoseNode]:
        if len(nodes) < 2:
            return list(nodes)
        if not edges and not gnss_priors:
            return list(nodes)

        graph = NonlinearFactorGraph()
        initial = Values()

        for node in nodes:
            initial.insert(node.index, Pose2(node.x, node.y, node.yaw))

        # 最初のノードをアンカー固定（グラフのゲージ自由度を除去する）
        anchor = nodes[0]
        prior_noise = noiseModel.Diagonal.Variances(_ANCHOR_VARIANCES)
        graph.add(PriorFactorPose2(
            anchor.index,
            Pose2(anchor.x, anchor.y, anchor.yaw),
            prior_noise,
        ))

        for edge in edges:
            noise = noiseModel.Gaussian.Information(edge.information)
            graph.add(BetweenFactorPose2(
                edge.from_index,
                edge.to_index,
                Pose2(edge.dx, edge.dy, edge.dyaw),
                noise,
            ))

        # GNSS絶対位置拘束を PoseTranslationPrior2D として投入する（xy平面のみ直接拘束）
        for gnss_prior in gnss_priors:
            if not initial.exists(gnss_prior.node_index):
                _logger.warning(
                    f'GNSS prior skipped: node_index={gnss_prior.node_index} not in graph'
                )
                continue
            gnss_noise = noiseModel.Gaussian.Information(gnss_prior.information)
            graph.add(PoseTranslationPrior2D(
                gnss_prior.node_index,
                Pose2(gnss_prior.x, gnss_prior.y, 0.0),
                gnss_noise,
            ))

        params = LevenbergMarquardtParams()
        params.setVerbosity('SILENT')
        result = LevenbergMarquardtOptimizer(graph, initial, params).optimize()

        updated: list[PoseNode] = []
        for node in nodes:
            pose = result.atPose2(node.index)
            updated.append(PoseNode(
                index=node.index,
                timestamp=node.timestamp,
                x=pose.x(),
                y=pose.y(),
                yaw=pose.theta(),
                scan=node.scan,
            ))

        _logger.info(
            f'GTSAMOptimizer: {len(nodes)} nodes, {len(edges)} edges, '
            f'{len(list(gnss_priors))} gnss_priors optimized'
        )
        return updated

File: slam_gnss_2d/slam_gnss_2d/src/optimizer/gtsam_optimizer.cpp
#include "slam_gnss_2d/optimizer/gtsam_optimizer.hpp"

#include <gtsam/geometry/Pose2.h>
#include <gtsam/linear/NoiseModel.h>
#include <gtsam/nonlinear/LevenbergMarquardtOptimizer.h>
#include <gtsam/nonlinear/LevenbergMarquardtParams.h>
#include <gtsam/nonlinear/NonlinearFactorGraph.h>
#include <gtsam/nonlinear/Values.h>
#include <gtsam/slam/BetweenFactor.h>
#include <gtsam/slam/PoseTranslationPrior.h>
#include <gtsam/slam/PriorFactor.h>

#include <rclcpp/rclcpp.hpp>

namespace slam_gnss_2d {
namespace optimizer {

namespace {
const gtsam::Vector3 kAnchorVariances(1e-6, 1e-6, 1e-8);
}

std::vector<core::PoseNode> GTSAMOptimizer::optimize(
    const std::vector<core::PoseNode>& nodes,
    const std::vector<core::PoseEdge>& edges,
    const std::vector<core::GnssPrior>& gnss_priors) {
  if (nodes.size() < 2) {
    return nodes;
  }
  if (edges.empty() && gnss_priors.empty()) {
    return nodes;
  }

  gtsam::NonlinearFactorGraph graph;
  gtsam::Values initial;

  for (const auto& node : nodes) {
    initial.insert(node.index, gtsam::Pose2(node.x, node.y, node.yaw));
  }

  const auto& anchor = nodes.front();
  auto prior_noise = gtsam::noiseModel::Diagonal::Variances(kAnchorVariances);
  graph.add(
      gtsam::PriorFactor<gtsam::Pose2>(
          anchor.index,
          gtsam::Pose2(anchor.x, anchor.y, anchor.yaw),
          prior_noise));

  for (const auto& edge : edges) {
    auto noise = gtsam::noiseModel::Gaussian::Information(edge.information);
    graph.add(
        gtsam::BetweenFactor<gtsam::Pose2>(
            edge.from_index,
            edge.to_index,
            gtsam::Pose2(edge.dx, edge.dy, edge.dyaw),
            noise));
  }

  for (const auto& gnss_prior : gnss_priors) {
    if (!initial.exists(gnss_prior.node_index)) {
      RCLCPP_WARN(
          rclcpp::get_logger("slam_gnss_2d.gtsam_optimizer"),
          "GNSS prior skipped: node_index=%d not in graph",
          gnss_prior.node_index);
      continue;
    }
    auto gnss_noise = gtsam::noiseModel::Gaussian::Information(gnss_prior.information);
    graph.add(
        gtsam::PoseTranslationPrior<gtsam::Pose2>(
            gnss_prior.node_index,
            gtsam::Point2(gnss_prior.x, gnss_prior.y),
            gnss_noise));
  }

  gtsam::LevenbergMarquardtParams params;
  params.setVerbosity("SILENT");
  gtsam::Values result = gtsam::LevenbergMarquardtOptimizer(graph, initial, params).optimize();

  std::vector<core::PoseNode> updated;
  updated.reserve(nodes.size());
  for (const auto& node : nodes) {
    const auto& pose = result.at<gtsam::Pose2>(node.index);
    updated.emplace_back(core::PoseNode{
        node.index,
        node.timestamp,
        pose.x(),
        pose.y(),
        pose.theta(),
        node.scan,
    });
  }

  RCLCPP_INFO(
      rclcpp::get_logger("slam_gnss_2d.gtsam_optimizer"),
      "GTSAMOptimizer: %zu nodes, %zu edges, %zu gnss_priors optimized",
      nodes.size(), edges.size(), gnss_priors.size());

  return updated;
}

}  // namespace optimizer
}  // namespace slam_gnss_2d
